sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TransformExecutorServices.java

```diff
@@ -0,0 +1,147 @@
+  }
+
+  /**
+   * Returns an EvaluationState that evaluates {@link TransformExecutor TransformExecutors} in
+   * serial.
+   */
+  public static TransformExecutorService serial(
+      ExecutorService executor, Map<TransformExecutor<?>, Boolean> scheduled) {
+    return new SerialEvaluationState(executor, scheduled);
+  }
+
+  /**
+   * The evaluation of a step where the input is unkeyed. Unkeyed inputs can be evaluated in
+   * parallel with an arbitrary amount of parallelism.
+   */
+  private static class ParallelEvaluationState implements TransformExecutorService {
+    private final ExecutorService executor;
+    private final Map<TransformExecutor<?>, Boolean> scheduled;
+
+    private ParallelEvaluationState(
+        ExecutorService executor, Map<TransformExecutor<?>, Boolean> scheduled) {
+      this.executor = executor;
+      this.scheduled = scheduled;
+    }
+
+    @Override
+    public void schedule(TransformExecutor<?> work) {
+      executor.submit(work);
+      scheduled.put(work, true);
+    }
+
+    @Override
+    public void complete(TransformExecutor<?> completed) {
+      scheduled.remove(completed);
+    }
+  }
+
+  /**
+   * The evaluation of a (Step, Key) pair. (Step, Key) computations are processed serially;
+   * scheduling a computation will add it to the Work Queue if a computation is in progress, and
+   * completing a computation will schedule the next item in the work queue, if it exists.
+   */
+  private static class SerialEvaluationState implements TransformExecutorService {
+    private final ExecutorService executor;
+    private final Map<TransformExecutor<?>, Boolean> scheduled;
+
+    private AtomicReference<TransformExecutor<?>> currentlyEvaluating;
+    private final Queue<TransformExecutor<?>> workQueue;
+
+    private SerialEvaluationState(
+        ExecutorService executor, Map<TransformExecutor<?>, Boolean> scheduled) {
+      this.scheduled = scheduled;
+      this.executor = executor;
+      this.currentlyEvaluating = new AtomicReference<>();
+      this.workQueue = new ConcurrentLinkedQueue<>();
+    }
+
+    /**
+     * Schedules the work, adding it to the work queue if there is a bundle currently being
+     * evaluated and scheduling it immediately otherwise.
+     */
+    @Override
+    public void schedule(TransformExecutor<?> work) {
+      workQueue.offer(work);
+      updateCurrentlyEvaluating();
+    }
+
+    @Override
+    public void complete(TransformExecutor<?> completed) {
+      if (!currentlyEvaluating.compareAndSet(completed, null)) {
+        throw new IllegalStateException(
+            "Finished work "
+                + completed
+                + " but could not complete due to unexpected currently executing "
+                + currentlyEvaluating.get());
+      }
+      scheduled.remove(completed);
+      updateCurrentlyEvaluating();
+    }
+
+    private void updateCurrentlyEvaluating() {
+      if (currentlyEvaluating.get() == null) {
```

Comment:
Suppose `A` and `B` are already in the queue, `C` is evaluating, and `D` is being scheduled.

I find this sequence odd:

1. `C` finishes and sets currently evaluating to `null`
2. `D`'s call to `updateCurrentlyEvaluating` sees there is nothing currently evaluating and enters critical section
3. `C`'s call to `updateCurrentlyEvaluating` blocks on critical section
4. `D`'s call schedules `A` which was already on the queue
5. `C`'s call is unblocked, dequeues `B`, compare and set fails, pushes `B` on the queue behind `D`.

Probably benign, but with a pathological schedule could probably starve a task. I've been shuffling interleavings of these methods in my head for a while, suspecting some danger of an actual atomicity problem. Is it a known performance issue if you have a more simple synchronization scheme?